
package samoa.core.protobuf;

enum CommandType {

    // Present only in SamoaResponse
    ERROR = 5;

    PING = 1;
    SHUTDOWN = 2;
    CLUSTER_STATE = 6;

    DECLARE_TABLE = 7;
    DROP_TABLE = 8;

    CREATE_PARTITION = 9;
    DROP_PARTITION = 10;

    GET_BLOB = 11;
    PUT_BLOB = 12;
};

// Returned by Samoa to indicate an error in the operation
message Error {
    required string type = 1;
    optional string message = 2;
};

message ClusterState {

    // A ClusterState represents a consistent snapshot of the cluster
    //  shape and configuration, from the perspective of a cluster
    //  participant.
    //
    // ClusterState messages are periodically passed between samoa
    //  servers for purposes of discovery: to detect unknown peers,
    //  tables, and partitions which should be tracked.

    required string local_uuid = 1;
    required string local_hostname = 2;
    required int32  local_port = 3;

    message Peer {

        // cluster-shared peer state
        required string uuid = 1;
        required string hostname = 2;
        required int32  port = 3;

        // local server statistics
        required bool   connected = 6;
        optional uint32 queue_size = 7;
        optional uint32 latency_ms = 8;
    };
    repeated Peer peer = 4;

    message Table {

        required string uuid = 1;
        optional bool   dropped = 2 [default = false];

        optional string name = 3;
        optional int32  data_type = 8;
        optional int64  lamport_consistency_bound = 9;
        optional int64  replication_factor = 4;


        message Partition {

            required string uuid = 1;
            optional bool   dropped = 2 [default = false];

            optional string server_uuid = 3;
            optional int64  ring_position = 4;
            optional int64  consistent_range_begin = 8;
            optional int64  consistent_range_end = 9;
            optional int64  lamport_ts = 10;
            optional string storage_path = 5;
            optional int64  storage_size = 6;
            optional int64  index_size = 7;
        };
        repeated Partition partition = 7;
    };
    repeated Table table = 5;
};

// Operations

message PingRequest {
    required int64 value_len = 1 [default = 0];
};
message PingResponse {
    required int64 value_len = 1 [default = 0];
};

// Table Manipulation

message DeclareTableRequest {
    required string name = 1;
    required string data_type = 2;
    required int32 lamport_consistency_bound = 3;
    required bool create_if_not_exists = 4 [default = true];
    required int32 replication_factor = 5 [default = 1];
};

message DeclareTableResponse {
    required string uuid = 1;
    required string name = 2;
    required string data_type = 3;
    required int32 lamport_consistency_bound = 4;
    required int32 replication_factor = 5;
    required bool created = 6;
};

message DropTableRequest {
    required string uuid = 1;
};

message DropTableResponse {
};

// Partition Manipulation

message CreatePartitionRequest {
    required string table_uuid = 1;
    required int32  ring_position = 2;

    required int64 storage_size = 3;
    required int64  index_size = 4;
};

message CreatePartitionResponse {
    required string uuid = 1;
};

message DropPartitionRequest {
    required string table_uuid = 1;
    required string uuid = 2;
};

message DropPartitionResponse {
};

// BLOB datatype operations

message GetBlobRequest {
    required string table_uuid = 1;
    required bytes key = 2;
};

message GetBlobResponse {
    required bool found = 1;
    optional int32 value_length = 2;
    optional bytes version_tag = 3;
};

message PutBlobRequest {
    required string table_uuid = 1;
    required bytes key = 2;
    required int32 value_length = 3;
    optional bytes version_tag = 4;
};

message PutBlobResponse {
    required bool success = 1;
    optional bytes version_tag = 2;
    optional int32 value_length = 3;
};


message SamoaRequest {
    required CommandType type = 1;

    // whether the client will be closing the connection after this request
    optional bool closing = 3 [default = false];

    optional PingRequest      ping = 2;
    optional ClusterState     cluster_state = 7;

    optional DeclareTableRequest declare_table = 8;
    optional DropTableRequest    drop_table = 9;

    optional CreatePartitionRequest create_partition = 10;
    optional DropPartitionRequest   drop_partition = 11;

    optional GetBlobRequest get_blob = 12;
    optional PutBlobRequest put_blob = 13;

    // generates an error response
    optional Error            error = 6;
};

message SamoaResponse {
    required CommandType type = 1;

    // whether the server will be closing the connection after this request
    optional bool closing = 6 [default = false];

    optional Error         error = 2;
    optional PingResponse  ping = 3;
    optional ClusterState  cluster_state = 7;

    optional DeclareTableResponse declare_table = 8;
    optional DropTableResponse    drop_table = 9;

    optional CreatePartitionResponse create_partition = 10;
    optional DropPartitionResponse   drop_partition = 11;

    optional GetBlobResponse get_blob = 12;
    optional PutBlobResponse put_blob = 13;
};

